### ISSUE TITLE
optimizer：When filter is false, select and select ... for update execute differently

### ISSUE THREAD

**Issue by Heoric** [cockroachdb/cockroach#73074]
I executed the following two queries and got different results,I think when the filter is false, it is not necessary to perform a full scan.

```sql
root@localhost:26257/defaultdb> explain select * from oracle where a =1 and a = 2;
         info
-----------------------
  distribution: local
  vectorized: true

  • norows
(4 rows)


Time: 1ms total (execution 1ms / network 0ms)

root@localhost:26257/defaultdb> explain select * from oracle where a =1 and a = 2 for update;
                                      info
---------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • filter
  │ estimated row count: 0
  │ filter: false
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 4 hours ago)
        table: oracle@oracle_pkey
        spans: FULL SCAN
        locking strength: for update
(12 rows)


Time: 2ms total (execution 1ms / network 0ms)
```

When I was debugging the code I found that the query with for update will set the Volatility `VolatilityVolatile`.

I think when the filter is false, the result should be stable.

```go
// logical_props_builder.go
	if scan.Locking != nil {
		rel.VolatilitySet.AddVolatile()
	}

// factory.go
if relational.Cardinality.IsZero() && relational.VolatilitySet.IsLeakProof() {
			if f.matchedRule == nil || f.matchedRule(opt.SimplifyZeroCardinalityGroup) {
				values := f.funcs.ConstructEmptyValues(relational.OutputCols)
				if f.appliedRule != nil {
					f.appliedRule(opt.SimplifyZeroCardinalityGroup, nil, values)
				}
				return values
			}
	}
```

Jira issue: CRDB-11408

**Comment by rytaft** [comment]
Hi @Heoric thanks for the report -- do you have a real-world query example that needs this?

**Comment by Heoric** [comment]
Yes, this kind of query is automatically generated by the application。

**Comment by rytaft** [comment]
Thanks for clarifying. It doesn't seem too difficult to make an exception for this case. We probably won't be able to prioritize this issue unless we see more users running into the problem, but I've marked it with a "quick win" tag so someone may pick it up during a fixit day.

**Comment by n0gu** [comment]
Hi, we also stumbled upon this issue, so leaving a note about our use case to +1 this. 
In our case, the query takes the form of `... WHERE column IN ($array) FOR UPDATE`, with $array sometimes being empty. I believe this is not an extremely rare case.

**Comment by michae2** [comment]
This should be fixed by the SFU changes in the v23.2 release.

**Comment by michae2** [comment]
This is improved by the new SFU implementation in 23.2 (`optimizer_use_lock_op_for_serializable `):

```
demo@127.0.0.1:26257/demoapp/defaultdb> CREATE TABLE oracle (a INT PRIMARY KEY);
CREATE TABLE

Time: 3ms total (execution 3ms / network 0ms)

demo@127.0.0.1:26257/demoapp/defaultdb> EXPLAIN SELECT * FROM oracle WHERE a = 1 AND a = 2;
         info
-----------------------
  distribution: local
  vectorized: true

  • norows
(4 rows)

Time: 7ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/demoapp/defaultdb> EXPLAIN SELECT * FROM oracle WHERE a = 1 AND a = 2 FOR UPDATE;
                 info
--------------------------------------
  distribution: local
  vectorized: true

  • filter
  │ estimated row count: 0
  │ filter: false
  │
  └── • scan
        missing stats
        table: oracle@oracle_pkey
        spans: FULL SCAN
        locking strength: for update
(12 rows)

Time: 1ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/demoapp/defaultdb> SET optimizer_use_lock_op_for_serializable = true;
SET

Time: 1ms total (execution 0ms / network 0ms)

demo@127.0.0.1:26257/demoapp/defaultdb> EXPLAIN SELECT * FROM oracle WHERE a = 1 AND a = 2 FOR UPDATE;
               info
----------------------------------
  distribution: local
  vectorized: true

  • lookup join (semi)
  │ estimated row count: 0
  │ table: oracle@oracle_pkey
  │ equality: (a) = (a)
  │ equality cols are key
  │ locking strength: for update
  │
  └── • norows
(11 rows)

Time: 1ms total (execution 1ms / network 0ms)
```

(I have filed https://github.com/cockroachdb/cockroach/issues/114751 about optimizing this further to simply norows.)
